### ISSUE TITLE
image and line functions aren't working on first page of a pdf

### ISSUE THREAD

**Issue by BramG** [dompdf/dompdf#1365]
I had to upgrade dompdf due to server changes.
At this moment it 'kinda works' again but the logo and some stylish line isn't appearing anymore on the first page. I'm using the latest version (0.7.0)

To summarise:
- **pageText** is working and printed on each screen
- **line** is working but NOT printed on the first screen
- **image** is working but NOT printed on the first screen

I also tried to use 'odd' as $where parameter but then nothing gets printed.
Any help would be appreciated.

```
  $pdf = new \Dompdf\Dompdf();

  $options = $pdf->getOptions();
  $options->set('isRemoteEnabled', TRUE);
  $options->set('isPhpEnabled', TRUE);
  $context = stream_context_create([
    'ssl' => [
      'verify_peer' => FALSE,
      'verify_peer_name' => FALSE,
      'allow_self_signed'=> TRUE
    ]
  ]);
  $pdf->setHttpContext($context);

  $pdf->set_base_path(realpath(path_to_theme() . '/'));
  $pdf->load_html($html);

  $pdf->render();

  $canvas = $pdf->getCanvas();

  $font_metric = new \Dompdf\FontMetrics($canvas, $options);
  $font = $font_metric->get_font('helvetica');
  $fontBold = $font_metric->get_font('helvetica', 'bold');

  // Header
  $path = drupal_get_path('theme', 'spock');
  $image = file_create_url($path . '/img/logo-pdf.png');

  $header = $canvas->open_object();
  $canvas->image($image, 472, 35, 85, 25);
  $canvas->line(0, 71, 458, 71, array(0, 0, 0), 5);
  $canvas->page_text(473, 66, t('specifications sheet'), $fontBold, 8, array(0, 0, 0));
  $canvas->close_object();
  $canvas->add_object($header, 'all');

  // Footer
  $footer = $canvas->open_object();
  $date = date('d-m-Y H:i');
  $canvas->page_text(35, 750, t($date . '   |   page {PAGE_NUM}/{PAGE_COUNT}'), $font, 7, array(0, 0, 0));
  $canvas->close_object();
  $canvas->add_object($footer, 'all');

  $pdf->stream($kr_number . '.pdf');
```

**Comment by simonberger** [comment]
Does something change if you remove writing the page_text and/or doing it before the line/image call?

**Comment by BramG** [comment]
Same result, if the pdf is only one page high it works perfectly
I simplified the example.

I also tested by removing the custom css, that makes it print on 3 pages.
In that case the logo is only printed on page 3.

So the problem is not that it isn't printed on the first page, the problem is that it is only printed on the last page.

  ```
$pdf = new \Dompdf\Dompdf();

  $options =$pdf->getOptions();

  $options->set('isRemoteEnabled', TRUE);
  $options->setIsPhpEnabled(TRUE);
  $contxt = stream_context_create([
    'ssl' => [
      'verify_peer' => FALSE,
      'verify_peer_name' => FALSE,
      'allow_self_signed'=> TRUE
    ]
  ]);
  $pdf->setHttpContext($contxt);

  $pdf->set_base_path(realpath(path_to_theme() . '/'));
  $pdf->load_html($html);

  $pdf->render();

  $canvas = $pdf->getCanvas();

  // Header
  $path = drupal_get_path('theme', 'spock');
  $image = file_create_url($path . '/img/logo-pdf.png');

  $header = $canvas->open_object();
  $canvas->image($image, 472, 35, 85, 25);
  $canvas->close_object();
  $canvas->add_object($header, 'all');
```

**Comment by simonberger** [comment]
I am surprised that should have worked at a point
The base_text explicitly says it can be added after rendering.
I am not sure that belongs to objects too. Also out of process operation view that makes no sense. 

Try to put the render call under your object bindings and there is a better chance it could work,

**Comment by BramG** [comment]
That's even worse, cause the objects don't get rendered when we do that.
The docs even state the canvas manipulations should happen after rendering.

Is there an alternative to achieve the same goal? Print an image and a line on each page?

It needs to be pixel perfect.

**Comment by BramG** [comment]
With the help of http://stackoverflow.com/questions/17143149/dompdf-draw-line-on-every-page

I found this solution:

```
  // Dirty fix for images and lines
  $canvas->page_script('
    $path = drupal_get_path(\'theme\', \'spock\');
    $image = file_create_url($path . \'/img/logo-pdf.png\');
    // $pdf is the variable containing a reference to the canvas object provided by dompdf
    $pdf->image($image, 472, 35, 85, 25);
    $pdf->line(0, 71, 458, 71, array(0, 0, 0), 5);
  ');
```

Seems dirty, but it works...

**Comment by bsweeney** [comment]
Let me illuminate the problem you're having. There is a bug in the current implementation where the canvas object that's used for the PDF (*) is not instantiated until the render method is called. If it were instantiated sooner you would be able to move your code above the render method and affect all pages. As it is, your code affects only the current page and any following pages. Because after render the canvas object is referencing the last page the only page affected is the last page.

Also, note that the page_text and page_script methods behave differently than the other methods (text, line, rectangle, etc). Other methods act only on the current page or object. The page_* methods are applied to each page during the packaging up of the PDF (i.e. during the stream or output methods). So the solution you found is not really a dirty method, just not ideal for what you're trying to do.

---

(*) I worded this statement in this way because the canvas is actually instantiated in the dompdf constructor, but it's re-instantiated during render. I know how to fix the bug, but haven't had time yet to actually implement it.

**Comment by BramG** [comment]
Thnx for the clear explanation! Really appreciated.

**Comment by simonberger** [comment]
An interesting task to understand the dompdf functionality a bit more.
Had a look at the problem finding out what really stops the object association working before the render call.
It normally would work. But the render call overwrites all changes previously done to the canvas.

I removed this reinitializing and the line from BramGs example got drawn to every page of my demo without breaking its CSS or anything else.
In this [commit ](https://github.com/simonberger/dompdf/commit/dbcfb3954f84a89f3910da541d008c7fdfa5db65) I added a compatible implementation skipping this dynamic page setup calculation by parameter (could be an option as well).

Just because this worked in my small test surely doesn't mean it can be done that easy and the options are covering all page setup options being calculated in the CSS method. I also didn't test what happens if the CSS method calculate a different result as being set in the options before (doing this test next).
Maybe we also "just" have to skip calculating the page setup in the css method?

**Comment by bsweeney** [comment]
The reason this canvas object is recreated in the render method is specifically to support specifying the page size in CSS. Removing the creation from here breaks that for sure.

Fixing this up so that we can refuting the page size should actually not be terribly difficult. It requires changes to the CPDF library and to the canvas class. I'm sure pdflib supports this functionality, but I haven't looked into it yet.

**Comment by simonberger** [comment]
Yeah I understand why the recreation exists. The commit doesn't break compatibility. The developer knows if he needs the CSS definition of page size. The commit opens the possibility to skip this recreation. It's needed for the least of us I bet. But still the skipping isn't per default to not break it.
This would be an easy change and could also be done without a parameter/option if no CSS page setup property has been found.

Anyway maybe you are right and its smarter to just change the canvas in a bigger change. I have no idea beside the canvas changes in which way that would affect earlier canvas additions/manipulations and so on.

**Comment by simonberger** [comment]
I do like it the mentioned alternative way.
Skipping the reinitializing if there is no need to, because the page size matches the default one.

[Committed here](https://github.com/dompdf/dompdf/commit/832073163fd013cf5ba54c75bdb90f084e768bfc)

In short it just changes

```php
if (is_array($basePageStyle->size)) {
   $this->setPaper(array(0, 0, $basePageStyle->size[0], $basePageStyle->size[1]));
}

//TODO: We really shouldn't be doing this; properties were already set in the constructor. We should add Canvas methods to set the page size and orientation after instantiaion (see #1059).
$this->setCanvas(CanvasFactory::get_instance($this, $this->paperSize, $this->paperOrientation));
$this->fontMetrics->setCanvas($this->pdf);
```

to

```php
if (is_array($basePageStyle->size)) {
   $basPageStyleSize = $basePageStyle->size;
   $optionPaperSize = $this->getPaperSize($options->getDefaultPaperSize());

   if (($basPageStyleSize[0] !== $optionPaperSize[0] || $basPageStyleSize[1] !== $optionPaperSize[1])) {
      $this->setPaper(array(0, 0, $basPageStyleSize[0], $basPageStyleSize[1]));
      $this->setCanvas(CanvasFactory::get_instance($this, $this->paperSize, $this->paperOrientation));
      $this->fontMetrics->setCanvas($this->pdf);
   }
}
```

It seems too easy. I racked my brain why that has been left out ages ago when this has been created. But I can't find another property changing the way the canvas will be created on the second run.
I came to the conclusion that it just has been overseen or didn't be payed much attention to at this point.

If this savely works it would also be a performance optimization beside the easier possible canvas manipulations.

**Comment by bsweeney** [comment]
OK, yes. I think that's a good temporary fix for the problem. There's certainly no need to re-create the canvas object if the page size doesn't change after parsing the styles.
